runtime/expr/agg/union.go

```diff
@@ -6,36 +6,33 @@ import (
 )
 
 type Union struct {
-	typ  zed.Type
-	val  map[string]struct{}
-	size int
+	types map[zed.Type]map[string]struct{}
+	size  int
 }
 
 var _ Function = (*Union)(nil)
 
 func newUnion() *Union {
 	return &Union{
-		val: make(map[string]struct{}),
+		types: make(map[zed.Type]map[string]struct{}),
 	}
 }
 
 func (u *Union) Consume(val *zed.Value) {
 	if val.IsNull() {
 		return
 	}
-	if u.typ == nil {
-		u.typ = val.Type
-	} else if u.typ != val.Type {
-		// We should make union type for the set-union
-		// instead of silently ignoring.  See #3363.
-		return
-	}
-	u.update(val.Bytes)
+	u.update(val.Type, val.Bytes)
 }
 
-func (u *Union) update(b zcode.Bytes) {
-	if _, ok := u.val[string(b)]; !ok {
-		u.val[string(b)] = struct{}{}
+func (u *Union) update(typ zed.Type, b zcode.Bytes) {
+	m, ok := u.types[typ]
+	if !ok {
+		m = make(map[string]struct{})
+		u.types[typ] = m
+	}
+	if _, ok := m[string(b)]; !ok {
+		m[string(b)] = struct{}{}
```

Comment:
Since `string([]byte{})` and `string([]byte(nil))` both evaluate to `""`, this won't distinguish between zero-length and nil values of `b`. One way to deal with that would be to change `Union.types` to `map[zed.Type]map[string]bool` and then store true if the `string` is `""` but represents a nil value.

Reply:
Actually this is moot since null values are ignored by the union aggregator.

Reply:
Excellent point! Sorry for missing that.